pkg/edit/prompt.go

```diff
@@ -126,6 +127,67 @@ func defaultStaleTransform(original ui.Text) ui.Text {
 	return ui.StyleText(original, ui.Inverse)
 }
 
+type rawToken struct {
+	data      []byte
+	isSGRCode bool
+}
+
+func parseBytesIntoTokens(b []byte) (ret []rawToken) {
+	currentSGRCode := []byte{}
+	currentPlaintext := []byte{}
+	isSGRSequence := false
+	skipCounter := 0
+	for idx, byt := range b {
+		if skipCounter > 0 {
+			skipCounter--
+			continue
+		}
+		if isSGRSequence {
+			if byt == 'm' {
+				currentSGRCode = append(currentSGRCode, 'm')
+				ret = append(ret, rawToken{
+					data:      currentSGRCode,
+					isSGRCode: true,
+				})
+				currentPlaintext = []byte{}
+				isSGRSequence = false
+				continue
+			}
+			currentSGRCode = append(currentSGRCode, byt)
+			continue
+		}
+		if len(b) > idx+1 {
+			byt2 := b[idx+1]
+			if byt == '\033' && byt2 == '[' {
+				skipCounter = 1
+				isSGRSequence = true
+				ret = append(ret, rawToken{
+					data:      currentPlaintext,
+					isSGRCode: false,
+				})
+				currentSGRCode = []byte{}
+				currentSGRCode = append(currentSGRCode, '\033', '[')
+				continue
+			}
+		}
+		currentPlaintext = append(currentPlaintext, byt)
+	}
+	return
+}
```

Comment:
Instead of parsing everything into a token list, I prefer implementing a stateful tokenizer type that produces one token at a time:

```go
func (st sgrTokenizer) next() sgrToken {
}
```

This method is about as easy to use as a list of tokens, but the implementation will be simpler - you can branch into an is-sgr-token branch vs. not-sgr-token branch right away, and avoid the state machine bookkeeping needed in the current implementation.

Please also take care to remove all escape characters that do not form valid SGR sequences.

Reply:
I would vote for leaving all escape characters unchanged. This would make it possible to emit custom escape sequences from the prompt function, for example to enable Terminal-specific functionality like iTerm2's Shell Integration. Example: https://github.com/zzamboni/elvish-modules/blob/master/iterm2.org

Reply:
@zzamboni, It is not clear why your proposal is needed. Doesn't the solution for issue #1039 address your use-case in a simpler, clearer, fashion? Also, this block of code does not, AFAIK, handle arbitrary escape sequences -- only SGR sequences which should not encompass the sort of terminal specific sequence you're thinking of.

Reply:
@krader1961 maybe I commented in the wrong place - I am not too familiar with the different types of escape sequences. What I know is what, at the moment, if I try to print some of iTerm2's [custom escape codes](https://www.iterm2.com/documentation-escape-codes.html) without redirecting them to `/dev/tty`, they get printed "escaped", like this:
![image](https://user-images.githubusercontent.com/32876/87130151-be2f3500-c292-11ea-8db8-8ad5e872c317.png)

As I understand, the proposal from #1039 would not change this if Elvish is still "escaping" those escape sequences instead of passing them unmodified.

Reply:
To clarify, and as mentioned in my latest comment in #1060, the problem with having to redirect with `/dev/tty` is that the codes are printed to the terminal when Elvish executes the prompt function to capture its output, and not when that output is actually printed to the screen. In the case of escape codes whose effect depends on where the cursor is when they are printed (like iTerm2's "end of prompt" marker), this makes a difference.

Reply:
@zzamboni, The solution for your situation is to implement a solution for issue #1039. Any escape sequences in a zero-width styled object are ignored by elvish -- as is any other text in such a segment. When you use a zero-width styled object you are promising that it does not change the location of the cursor. Thus there is no need for elvish to interpret the bytes within such a segment in an attempt to determine the width of the output produced by the segment.

Reply:
@krader1961 thanks for the explanation. I agree zero-width segments solve this problem. See my latest comment in #1060 - I still think having a post-prompt hook would be useful but agree that zero-width segments would be far more important.